src/AspNet.Security.OAuth.MailRu/MailRuAuthenticationHandler.cs

```diff
@@ -0,0 +1,131 @@
+                    .AddOptionalClaim(ClaimTypes.Name, MailRuAuthenticationHelper.GetNickname(payload), Options.ClaimsIssuer)
+                    .AddOptionalClaim(ClaimTypes.Surname, MailRuAuthenticationHelper.GetFamilyName(payload), Options.ClaimsIssuer)
+                    .AddOptionalClaim(ClaimTypes.GivenName, MailRuAuthenticationHelper.GetGivenName(payload), Options.ClaimsIssuer)
+                    .AddOptionalClaim(ClaimTypes.Webpage, MailRuAuthenticationHelper.GetLink(payload), Options.ClaimsIssuer)
+                    .AddOptionalClaim(ClaimTypes.Email, MailRuAuthenticationHelper.GetEmail(payload), Options.ClaimsIssuer);
+
+            var principal = new ClaimsPrincipal(identity);
+            var ticket = new AuthenticationTicket(principal, properties, Options.AuthenticationScheme);
+
+            var context = new OAuthCreatingTicketContext(ticket, Context, Options, Backchannel, tokens, payload);
+            await Options.Events.CreatingTicket(context);
+
+            return context.Ticket;
+        }
+
+        protected override async Task<OAuthTokenResponse> ExchangeCodeAsync([NotNull] string code, [NotNull] string redirectUri)
+        {
+            var credentials = Convert.ToBase64String(Encoding.UTF8.GetBytes($"{Options.ClientId}:{Options.ClientSecret}"));
+
+            var request = new HttpRequestMessage(HttpMethod.Post, Options.TokenEndpoint);
+            request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+            request.Headers.Authorization = new AuthenticationHeaderValue("Basic", credentials);
+
+            request.Content = new FormUrlEncodedContent(new Dictionary<string, string>
+            {
+                ["grant_type"] = "authorization_code",
+                ["redirect_uri"] = redirectUri,
+                ["code"] = code
+            });
+
+            var response = await Backchannel.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, Context.RequestAborted);
+            if (!response.IsSuccessStatusCode)
+            {
+                Logger.LogError("An error occurred while retrieving an access token: the remote server " +
+                                "returned a {Status} response with the following payload: {Headers} {Body}.",
+                                /* Status: */ response.StatusCode,
+                                /* Headers: */ response.Headers.ToString(),
+                                /* Body: */ await response.Content.ReadAsStringAsync());
+
+                return OAuthTokenResponse.Failed(new Exception("An error occurred while retrieving an access token."));
+            }
+
+            var payload = JObject.Parse(await response.Content.ReadAsStringAsync());
+
+            return OAuthTokenResponse.Success(payload);
+        }
+
+        protected virtual string BuildUrlWithSignature(string baseUrl, IDictionary<string, string> parameters, string signKey)
```

Comment:
Unfortunately, this method won't work if the user adds custom query string parameters to the `Options.UserInformationEndpoint` address. To fix that, you'll need to parse the address, just like what the Instagram provider does.